[Bug]: Index Out Of Range for gist.github.com

### What happened?

When I was preparing another goreleaser bug, I was using gist.github.com as a sample repository.
When I ran, then the following error occured:

```bash
$ goreleaser release --skip-publish --rm-dist
   • releasing...
   • could not find a config file, using defaults...
   • loading environment variables
   • getting and validating git state
      • building...               commit=9bd0f955cde0b405962e483a6c7b27770a91c956 latest tag=v1.0.0
   • parsing tag
   • running before hooks
   • setting defaults
      • snapshotting
      • github/gitlab/gitea releases
panic: runtime error: index out of range [-1]

goroutine 10 [running]:
github.com/goreleaser/goreleaser/internal/git.ExtractRepoFromURL(0xc00004b0e0, 0x30, 0x3, 0x5, 0xc0004ab2c0, 0x39)
	github.com/goreleaser/goreleaser/internal/git/config.go:38 +0x285
github.com/goreleaser/goreleaser/internal/git.ExtractRepoFromConfig(0x2110a20, 0x2fd7338, 0x21115a0, 0xc000535d10, 0xc0000a0100, 0x24c5d18)
	github.com/goreleaser/goreleaser/internal/git/config.go:20 +0x12b
github.com/goreleaser/goreleaser/internal/pipe/release.Pipe.Default(0xc000828400, 0x24, 0x0)
	github.com/goreleaser/goreleaser/internal/pipe/release/release.go:80 +0x48b
github.com/goreleaser/goreleaser/internal/middleware.ErrHandler.func1(0xc000828400, 0xc00004b020, 0x24)
	github.com/goreleaser/goreleaser/internal/middleware/error.go:13 +0x33
github.com/goreleaser/goreleaser/internal/middleware.Logging.func1(0xc000828400, 0x0, 0x0)
	github.com/goreleaser/goreleaser/internal/middleware/logging.go:33 +0x18e
github.com/goreleaser/goreleaser/internal/pipe/defaults.Pipe.Run(0xc000828400, 0x18, 0x0)
	github.com/goreleaser/goreleaser/internal/pipe/defaults/defaults.go:40 +0x1ca
github.com/goreleaser/goreleaser/internal/middleware.ErrHandler.func1(0xc000828400, 0xc000049128, 0x18)
	github.com/goreleaser/goreleaser/internal/middleware/error.go:13 +0x33
github.com/goreleaser/goreleaser/internal/middleware.Logging.func1(0xc000828400, 0x0, 0x0)
	github.com/goreleaser/goreleaser/internal/middleware/logging.go:33 +0x18e
github.com/goreleaser/goreleaser/cmd.releaseProject.func1(0x0, 0x0)
	github.com/goreleaser/goreleaser/cmd/release.go:108 +0x11b
github.com/caarlos0/ctrlc.(*Ctrlc).Run.func1(0xc00007e290, 0xc0005356a0)
	github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:39 +0x27
created by github.com/caarlos0/ctrlc.(*Ctrlc).Run
	github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:38 +0x65
```

and

```bash
$ goreleaser check
   • could not find a config file, using defaults...
   • checking config:
      • snapshotting
      • github/gitlab/gitea releases
panic: runtime error: index out of range [-1]

goroutine 25 [running]:
github.com/goreleaser/goreleaser/internal/git.ExtractRepoFromURL(0xc000178150, 0x30, 0x3, 0x5, 0xc0004d62c0, 0x39)
	github.com/goreleaser/goreleaser/internal/git/config.go:38 +0x285
github.com/goreleaser/goreleaser/internal/git.ExtractRepoFromConfig(0x2110a20, 0x2fd7338, 0x21115a0, 0xc000118f50, 0xc000138100, 0x24c5d18)
	github.com/goreleaser/goreleaser/internal/git/config.go:20 +0x12b
github.com/goreleaser/goreleaser/internal/pipe/release.Pipe.Default(0xc000502d00, 0x24, 0x0)
	github.com/goreleaser/goreleaser/internal/pipe/release/release.go:80 +0x48b
github.com/goreleaser/goreleaser/internal/middleware.ErrHandler.func1(0xc000502d00, 0xc000178090, 0x24)
	github.com/goreleaser/goreleaser/internal/middleware/error.go:13 +0x33
github.com/goreleaser/goreleaser/internal/middleware.Logging.func1(0xc000502d00, 0x0, 0x0)
	github.com/goreleaser/goreleaser/internal/middleware/logging.go:33 +0x18e
github.com/goreleaser/goreleaser/internal/pipe/defaults.Pipe.Run(0xc000502d00, 0xc0005a6018, 0x18)
	github.com/goreleaser/goreleaser/internal/pipe/defaults/defaults.go:40 +0x1ca
github.com/goreleaser/goreleaser/cmd.newCheckCmd.func1.1(0x0, 0x0)
	github.com/goreleaser/goreleaser/cmd/check.go:46 +0xe5
github.com/caarlos0/ctrlc.(*Ctrlc).Run.func1(0xc000118280, 0xc0003c3850)
	github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:39 +0x27
created by github.com/caarlos0/ctrlc.(*Ctrlc).Run
	github.com/caarlos0/ctrlc@v1.0.0/ctrlc.go:38 +0x65
```

### How can we reproduce this?

```bash
git clone https://gist.github.com/cimnine/eb61e0bbd63b509c21ff33058c408eee reproduce
cd reproduce
goreleaser release --skip-publish --rm-dist
```

### goreleaser version

```bash
$ goreleaser --version
goreleaser version 0.175.0
commit: 5e006eb236284c29143341e48c2f692826865d47
built by: homebrew

https://goreleaser.com
```


### GoReleaser Check

- [X] goreleaser check shows no errors

### Search

- [X] I did search for other open and closed issues before opening this.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

> If I run `goreleaser check` it shows no errors

`goreleaser check` panics as well, so the answer would actually be no.

> By submitting this issue, you agree to follow our Code of Conduct

The _Code of Conduct_ links to https://example.com .

--- a/internal/git/config.go
+++ b/internal/git/config.go
@@ -3,6 +3,7 @@ package git
 import (
 	"errors"
 	"fmt"
+	"net/url"
 	"strings"
 
 	"github.com/goreleaser/goreleaser/pkg/config"
@@ -17,25 +18,35 @@ func ExtractRepoFromConfig() (result config.Repo, err error) {
 	if err != nil {
 		return result, fmt.Errorf("repository doesn't have an `origin` remote")
 	}
-	return ExtractRepoFromURL(out), nil
+	return ExtractRepoFromURL(out)
 }
 
-func ExtractRepoFromURL(s string) config.Repo {
+func ExtractRepoFromURL(rawurl string) (config.Repo, error) {
 	// removes the .git suffix and any new lines
-	s = strings.NewReplacer(
-		".git", "",
-		"\n", "",
-	).Replace(s)
+	s := strings.TrimSuffix(strings.TrimSpace(rawurl), ".git")
+
 	// if the URL contains a :, indicating a SSH config,
 	// remove all chars until it, including itself
 	// on HTTP and HTTPS URLs it will remove the http(s): prefix,
 	// which is ok. On SSH URLs the whole user@server will be removed,
 	// which is required.
 	s = s[strings.LastIndex(s, ":")+1:]
-	// split by /, the last to parts should be the owner and name
-	ss := strings.Split(s, "/")
+
+	// now we can parse it with net/url
+	u, err := url.Parse(s)
+	if err != nil {
+		return config.Repo{}, err
+	}
+
+	// split the parsed url path by /, the last two parts should be the owner and name
+	ss := strings.Split(strings.TrimPrefix(u.Path, "/"), "/")
+
+	// if less than 2 parts, its likely not a valid repository
+	if len(ss) < 2 {
+		return config.Repo{}, fmt.Errorf("unsupported repository URL: %s", rawurl)
+	}
 	return config.Repo{
 		Owner: ss[len(ss)-2],
 		Name:  ss[len(ss)-1],
-	}
+	}, nil
 }

--- a/internal/git/config_test.go
+++ b/internal/git/config_test.go
@@ -18,6 +18,7 @@ func TestRepoName(t *testing.T) {
 }
 
 func TestExtractRepoFromURL(t *testing.T) {
+	// valid urls
 	for _, url := range []string{
 		"git@github.com:goreleaser/goreleaser.git",
 		"git@custom:goreleaser/goreleaser.git",
@@ -27,8 +28,21 @@ func TestExtractRepoFromURL(t *testing.T) {
 		"https://github.enterprise.com/crazy/url/goreleaser/goreleaser.git",
 	} {
 		t.Run(url, func(t *testing.T) {
-			repo := git.ExtractRepoFromURL(url)
+			repo, err := git.ExtractRepoFromURL(url)
+			require.NoError(t, err)
 			require.Equal(t, "goreleaser/goreleaser", repo.String())
+		})
+	}
+
+	// invalid urls
+	for _, url := range []string{
+		"git@gist.github.com:someid.git",
+		"https://gist.github.com/someid.git",
+	} {
+		t.Run(url, func(t *testing.T) {
+			repo, err := git.ExtractRepoFromURL(url)
+			require.EqualError(t, err, "unsupported repository URL: "+url)
+			require.Equal(t, "", repo.String())
 		})
 	}
 }

